File: java/StringDeletion.java
/*
*   Title: String deletion
*
*   Given a string and a dictionary HashSet, write a function to determine the 
*   minimum number of characters to delete to make a word.
*
*   Execution: javac StringDeletion.java && java StringDeletion
*
*   For more details, check out http://www.byte-by-byte.com/stringdeletion
*/
import java.util.*;

public class StringDeletion {
    public static int delete(String query, HashSet<String> dictionary) {
        Queue<String> queue = new LinkedList<String>();
        Set<String> queueElements = new HashSet<String>();

        queue.add(query);
        queueElements.add(query);

        while (!queue.isEmpty()) {
            String s = queue.remove();
            queueElements.remove(s);
            if (dictionary.contains(s)) {
               return query.length() - s.length();
            }
            for (int i = 0; i < s.length(); i++) { 
                String sub = s.substring(0, i) + s.substring(i+1, s.length()); 
                if (sub.length() > 0 && !queueElements.contains(sub)) {
                    queue.add(sub);
                    queueElements.add(sub);
                }
            }
        }
        return -1;
    }

    // Sample test cases
    public static void main(String[] args) {
        String query = "abc";
        HashSet<String> hset = new HashSet<String>();

        hset.add("a"); 
        hset.add("aa"); 
        hset.add("aaa");

        assert delete(query, hset) == 2:
            "Successfully determines 2 strings.";

        System.out.println("Passed all test cases");
    }
}

File: python/StringDeletion.py
"""
  Title: String deletion

  Given a string and a dictionary HashSet, write a function to determine the 
  minimum number of characters to delete to make a word.

  Execution: python StringDeletion.py

  For more details, check out http://www.byte-by-byte.com/stringdeletion
"""
from collections import deque
import unittest


def delete(query: str, dictionary: list):
    queue = deque()
    queue_elements = set()

    queue.append(query)
    queue_elements.add(query)

    while queue:
        s = queue.popleft()
        queue_elements.remove(s)
        if s in dictionary:
            return len(query) - len(s)
        for i in range(len(s)):
            sub = s[0:i] + s[i+1:len(s)]
            if len(sub) > 0 and sub not in queue_elements:
                queue.append(sub)
                queue_elements.add(sub)
    return -1


class TestStringDeletion(unittest.TestCase):
    def test_I(self):
        self.assertEqual(delete("abc", ["a", "aa", "aaa"]), 2)
        print("Successfully determines 2 strings.")


if __name__ == '__main__':
    unittest.main()



